### PR TITLE
Provide machine networks and cluster egress rules in firewall spec

### PR DESCRIPTION
**Pull request by mwindower**
- use a custom reconciler for egress ips so that they are reconciled in every case
- provide machine networks and egress rules with the firewall spec that is deployed to shoots
- signs the values passed to the firewall chart with the K8s CA private key; therewith the signature can easily be checked by the firewall-controller (it has already a kubeconfig with the CA's public key)

### CODE REVIEW

**Comment by mwindower** [comment]
RFC @Gerrit91 @majst01

File: pkg/apis/metal/v1alpha1/types_infrastructure.go [hunk 1]
```diff
@@ -15,10 +15,35 @@ type InfrastructureConfig struct {
...
+type MachineNetwork struct {
+	Asn                 *int64   `json:"asn"`
+	Destinationprefixes []string `json:"destinationprefixes"`
+	Ips                 []string `json:"ips"`
+	Nat                 *bool    `json:"nat"`
+	Networkid           *string  `json:"networkid"`
+	Networktype         *string  `json:"networktype"`
+	Prefixes            []string `json:"prefixes"`
+	Private             *bool    `json:"private"`
+	Underlay            *bool    `json:"underlay"`
```
Comment (Gerrit91) [inline pkg/apis/metal/v1alpha1/types_infrastructure.go:35]: Does it still need these fields? I thought they were deprecated.

File: pkg/apis/metal/v1alpha1/types_infrastructure.go [hunk 2]
```diff
@@ -15,10 +15,35 @@ type InfrastructureConfig struct {
...
 
 type Firewall struct {
-	Size       string            `json:"size"`
-	Image      string            `json:"image"`
-	Networks   []string          `json:"networks"`
-	RateLimits map[string]uint32 `json:"rateLimits"`
+	Size            string           `json:"size"`
+	Image           string           `json:"image"`
+	Networks        []string         `json:"networks"`
+	MachineNetworks []MachineNetwork `json:"machinenetworks"`
```
Comment (Gerrit91) [inline pkg/apis/metal/v1alpha1/types_infrastructure.go:21]: Does the user really need to specify these complex machine networks at shoot creation time? I thought this was something dynamic?

Comment (mwindower) [reply pkg/apis/metal/v1alpha1/types_infrastructure.go]: That's right - removed it.

File: pkg/apis/metal/validation/infrastructure.go [hunk 3]
```diff
@@ -93,9 +93,41 @@ func ValidateInfrastructureConfig(infra *apismetal.InfrastructureConfig) field.E
...
+			allErrs = append(allErrs, field.Required(fp, "egress rule network must not be an empty string"))
+			continue
+		}
+		if !availableNetworks.Has(egress.NetworkID) {
+			allErrs = append(allErrs, field.Required(fp, "egress rule network must be present as cluster network"))
+			continue
+		}
+		if len(egress.IPs) == 0 {
+			allErrs = append(allErrs, field.Required(fp, "egress rule must contain ip addresses to use"))
+			continue
```
Comment (Gerrit91) [inline pkg/apis/metal/validation/infrastructure.go:130]: Could be a nice verification to check if the IP address is parsable with `net.ParseIP`.

File: charts/internal/shoot-control-plane/templates/firewall/firewall.yaml [hunk 4]
```diff
@@ -11,11 +11,19 @@ metadata:
   name: firewall
```
Comment (Gerrit91) [inline charts/internal/shoot-control-plane/templates/firewall/firewall.yaml:11]: Would be nice if the values from the chart would be reflected in the chart's default values file.

File: pkg/controller/infrastructure/actuator_reconcile.go [hunk 5]
```diff
@@ -391,6 +411,112 @@ func createFirewall(ctx context.Context, r *firewallReconciler) error {
...
+
+			dbIP := resp.IPs[0]
+			if dbIP.Type != nil && *dbIP.Type != metalgo.IPTypeStatic {
+				return &controllererrors.RequeueAfterError{
+					Cause:        errors.Wrap(err, fmt.Sprintf("ips for egress rule must be static, but %s is not static", ip)),
+					RequeueAfter: 30 * time.Second,
+				}
+			}
+
+			if len(dbIP.Tags) > 0 {
```
Comment (Gerrit91) [inline pkg/controller/infrastructure/actuator_reconcile.go:469]: From the error message it says that the IP does not have an egress tag. Where is this check if this IP has this kind of tag or not? And if it already has one, couldn't the length of `dbIP.Tags` be `1` as well?

Comment (Gerrit91) [reply pkg/controller/infrastructure/actuator_reconcile.go:469]: Ah, I think I got it, it's the first condition where you `continue`. So, we will somewhere need to document that a user is not allowed to use tags on IPs when he wants to use the SNAT feature. Could be nice to allow custom user tags in the future, I guess.

**Review by Gerrit91 — APPROVED** [review]

File: pkg/controller/infrastructure/actuator_reconcile.go (hunk 5 above)
Comment (mwindower) [reply pkg/controller/infrastructure/actuator_reconcile.go:469]: If the IP has just a single egress tag for this cluster, it's already contained in `currentEgressIPs`: https://github.com/metal-stack/gardener-extension-provider-metal/blob/f62fb633a38d8075eee4c20168962be5e4eb6af9/pkg/controller/infrastructure/actuator_reconcile.go#L437
If the IP has some tags at the given line, it's assumed to be already taken for some other purpose.

**Comment by mwindower** [comment]
tampering protection needs to be implemented

File: charts/internal/shoot-control-plane/templates/firewall/firewall.yaml [hunk 6]
```diff
@@ -11,11 +11,21 @@ metadata:
...
+{{- if .Values.firewall.ratelimits }}
   ratelimits:
-{{ toYaml .Values.firewall.rateLimits | indent 4 }}
+{{ toYaml .Values.firewall.ratelimits | indent 4 }}
+{{- end }}
+{{- if .Values.firewall.egressrules }}
+  egressrules:
+{{ toYaml .Values.firewall.egressrules | indent 4 }}
 {{- end }}
+  signature:
```
Comment (mwindower) [inline charts/internal/shoot-control-plane/templates/firewall/firewall.yaml:30]: revert to camel case for chart value fields

File: pkg/apis/metal/v1alpha1/types_infrastructure.go [hunk 7]
```diff
@@ -15,10 +15,21 @@ type InfrastructureConfig struct {
...
 type Firewall struct {
-	Size       string            `json:"size"`
-	Image      string            `json:"image"`
-	Networks   []string          `json:"networks"`
-	RateLimits map[string]uint32 `json:"rateLimits"`
+	Size        string       `json:"size"`
+	Image       string       `json:"image"`
+	Networks    []string     `json:"networks"`
+	RateLimits  []RateLimit  `json:"ratelimits"`
+	EgressRules []EgressRule `json:"egressrules"`
```
Comment (mwindower) [inline pkg/apis/metal/v1alpha1/types_infrastructure.go:22]: use camel case

File: pkg/controller/controlplane/valuesprovider.go [hunk 8]
```diff
@@ -540,33 +543,63 @@ func (vp *valuesProvider) getControlPlaneShootChartValues(ctx context.Context, c
...
+	if err != nil {
+		return nil, errors.Wrap(err, "could not find firewall for cluster")
+	}
+	if len(firewalls) != 1 {
+		return nil, fmt.Errorf("cluster %s has %d firewalls", clusterID, len(firewalls))
+	}
+
+	firewall := *firewalls[0]
+	fwValues := map[string]interface{}{
+		"machinenetworks":  firewall.Allocation.Networks,
```
Comment (mwindower) [inline pkg/controller/controlplane/valuesprovider.go:574]: use explicit mapping from `V1MachineNetwork` of metal-api to own `MachineNetwork` struct

**Review by majst01 — APPROVED** [review]

**Review by Gerrit91 — APPROVED** [review]

**Pull request merged**